### PR TITLE
Add Debian packaging via .mk file

### PR DESCRIPTION
**Pull request by dpezely**
Add Debian packaging:
- Ticket: [hpr#16](https://github.com/helium/helium-packet-router/issues/16)
- Keeps all of the Debian-ish procedures isolated from primary `Makefile`
- Copies file from `rebar3 release` into debian directory tree via `rsync` using hard-links to avoid duplicating files
- Usage: `make deb`

TO DO:
- [x] construct Debian package `.deb` file based upon a successful `rebar3 release` run
- [x] confirmed .deb generation against `router` repo; see below
- [ ] confirmed .deb installation on clean AWS instance
- [ ] ensure Debian-compliant version numbering
- [ ] `systemd` config and launch scripts (required)
- [ ] `/etc/init/` and `/etc/rc?.d/` launch scripts (optional, but good for community)
- [ ] Successful submission to official Debian repo (probably after merging this PR)

APPENDIX:

Extra bits used for making .deb for [router](https://github.com/helium/router) follow.

File: `../router/Make-deb-package.mk`

```Makefile
NAME=router
all:
	NAME=${NAME} SHORT_NAME=${NAME} \
	  make -f ../helium-packet-router/Make-deb-package.mk
```

File: `../router/priv/debian/deb-control.txt`
```conf
Package: helium-router
Version: 1.0-1
Section: lorawan
Priority: extra
Architecture: all
Depends: 
Suggests: esl-erlang
Installed-Size: 47185920
Maintainer: Nova Labs Inc <support@nova-labs.com>
Conflicts: helium-router
Provides: helium-router
Description: LoRaWAN Network Server (LNS)
 conforming to LoRaWAN 1.0.x specification.
 See https://github.com/helium/router
```

Run:
```bash
cd ~/helium/router/
make -f Make-deb-package.mk
```

Once this PR gets merged, we can back-port to `router` repo.

### CODE REVIEW

File: Make-deb-package.mk [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	echo "${DEST}/bin/${NAME}.sh unkind-stop" >> release/DEBIAN/prerm
+	echo "${DEST}/bin/${NAME}.sh clean-logs" >> release/DEBIAN/prerm
+	echo "rm -f ${DEST}/var/log/${NAME}.log*" >> release/DEBIAN/prerm
+	echo "rm -f /etc/init.d/${NAME}.sh" >> release/DEBIAN/prerm
+	echo "rm -f /etc/rc?.d/*${NAME}" >> release/DEBIAN/prerm
+	echo 'userdel -r ${RUNTIME_USER}' >> release/DEBIAN/prerm
+
+	chmod 755 release/DEBIAN/p*
+
+	rsync -a --link-dest=$(shell pwd)/${REBAR_BUILD_DIR} \
```
Comment (jontow) [inline Make-deb-package.mk:98]: very optional idea: rsync being common but not always preinstalled, maybe we could check for its presence in the `prerelease` target and bail if not found, before spending time compiling anything? envisioning something like this:  `rsync --help >/dev/null 2>&1 || echo "rsync not installed or missing from PATH"`

File: Make-deb-package.mk [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	  ${REBAR_BUILD_DIR}/ release${DEST}/
+
+	dpkg-deb --build release
```
Comment (jontow) [inline Make-deb-package.mk:101]: According to the manpage, I think 101 and 102 can be combined here to avoid the mv(1):
`dpkg-deb --build release ${DEBIAN_PACKAGE_NAME}_${REVISION}_${DEB_ARCH}.deb`

**Review by jontow — DISMISSED** [review]
way nice first pass, added some possible improvement comments, but nothing that stops my :+1:

File: Make-deb-package.mk (hunk 2 above)
Comment (dpezely) [reply Make-deb-package.mk]: Nice!  Fixed in commit a544c77

File: Make-deb-package.mk (hunk 1 above)
Comment (dpezely) [reply Make-deb-package.mk:102]: I went with `which` instead of `--help` which should accomplish the same thing.

Fixed in commit a544c77

File: Make-deb-package.mk (hunk 2 above)
Comment (dpezely) [reply Make-deb-package.mk]: And that fix was corrected with commit f93c870

File: Make-deb-package.mk (hunk 1 above)
Comment (dpezely) [reply Make-deb-package.mk:103]: And that fix was corrected with commit f93c870

**Comment by dpezely** [comment]
This effort is **blocked** until Helium packet-router "main" branch can run `make rel` successfully.

**Comment by dpezely** [comment]
> This effort is **blocked** until Helium packet-router "main" branch can run `make rel` successfully.

@jadeallenx Rumor has it that you're versed in `relx`.  Would you be able to help drag this over the finish line for making a simple Debian package, please?

TL;DR: 
- Checklist in ticket is up-to-date: https://github.com/helium/helium-packet-router/issues/16
- Run `make deb` or for more nuanced commands, use: `make -f Make-deb-package.mk vm-test vm-ssh vm-relx`
- Background:
  + I thought `relx` bits were successful based upon `router background` *apparently* returning cleanly but silently exiting with `127`
  + Based upon that perception of the (false) positive launch, I wrote all the other things: systemd service, init.d & rc*.d scripts, Vagrantfile, etc.
  + `rebar3 relx` successfully creates something runnable within Docker
  + However, using `make -f Make-deb-package.mk vm-relx` which is a simple Vagrant VM wrapped around that same `rel` directory tree, nothing runs / always exits with status `127`.
  + Likewise, the resulting `.deb` file fails to launch the app during install

**Review by dpezely — COMMENTED** [review]
Excellent!  Thank you!

Since I created the initial PR, GitHub won't allow me to approve despite Jade overhauling and finishing it.

File: .github/scripts/make_deb.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+VERSION=$(git describe)
+
+DIAGNOSTIC=1 ./rebar3 release
```
Comment (michaeldjeffrey) [inline .github/scripts/make_deb.sh:7]: Do we need to use `make rel` here to make sure grpc deps are built?

File: deb/after_install.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# add miner to /usr/local/bin it appears in path
```
Comment (michaeldjeffrey) [inline deb/after_install.sh:1]:
```suggestion
# add hpr to /usr/local/bin it appears in path
```

File: deb/before_upgrade.sh [hunk 5]
```diff
@@ -0,0 +1,2 @@
+# add system user for file ownership and systemd user, if not exists
+useradd --system --home-dir /opt/miner --create-home helium || true
```
Comment (michaeldjeffrey) [inline deb/before_upgrade.sh:2]:
```suggestion
useradd --system --home-dir /opt/hpr --create-home helium || true
```

File: deb/after_upgrade.sh [hunk 6]
```diff
@@ -0,0 +1,2 @@
+# add miner to /usr/local/bin so it appears in path, if it does not already exist
```
Comment (michaeldjeffrey) [inline deb/after_upgrade.sh:1]:
```suggestion
# add hpr to /usr/local/bin so it appears in path, if it does not already exist
```

**Review by dpezely — COMMENTED** [review]
There's `config/sys.config.src`, but does this also need `config/sys.config` as well?

(I'm still baffled about why there would be two such files-- why, Erlang, why?)

File: rebar.config [hunk 7]
```diff
@@ -3,6 +3,7 @@
 {cover_enabled, true}.
 
 {deps, [
+    recon,
```
Comment (michaeldjeffrey) [inline rebar.config:6]:
```suggestion
```

We already have recon pinned to a version at the bottom of the deps.

**Comment by jadeallenx** [comment]
> (I'm still baffled about why there would be two such files-- why, Erlang, why?)

A `.src` file is a template which rebar uses to create the "real" sys.config at runtime - it is not natively supported by ERTS; the `sys.config` file is fhe old classic file which relx still supports. You need one or the other, not both, although there are plenty of projects which have both. `relx` prefers `.src` though and if it exists, will use it instead of plain ol' `sys.config.

**Review by michaeldjeffrey — DISMISSED** [review]

File: config/sys.config.src [hunk 8]
```diff
@@ -1,6 +1,6 @@
 [
     {hpr, [
-        {base_dir, "/var/data"}
+        {base_dir, "data"}
```
Comment (macpie) [inline config/sys.config.src:3]: base_dir os not needed anymore

File: rebar.config [hunk 9]
```diff
@@ -67,12 +67,12 @@
...
 
 {relx, [
-    {release, {hpr, "1.0.0"}, [hpr]},
+    {release, {hpr, git}, [hpr, runtime_tools, tools, recon]},
     {vm_args, "config/vm.args"},
-    {sys_config, "config/sys.config"},
     {sys_config_src, "config/sys.config.src"},
-    {include_src, true},
-    {overlay, []},
+    {include_src, false},
```
Comment (macpie) [inline rebar.config:73]: Does this mean we won't be able to hot load?

Comment (jadeallenx) [reply rebar.config:73]: Don't think so, just makes the image size as small as possible - by hot loading do you mean using sasl and systools like `relujp` or just sideloading new beams?

Comment (macpie) [reply rebar.config:73]: > sideloading new beams

**Review by macpie — APPROVED** [review]
